### PR TITLE
Merge `rxinference`and `inference` functions into `infer`

### PR DESCRIPTION
**Pull request by albertpod**
This PR addresses issue #181 and marks significant update to the `RxInfer`, wherein we merge the functionalities of `inference` and `rxinference` into a single, "umbrella" function: `infer`. This change simplifies the user interface by providing a unified method for conducting both static and streaming probabilistic inference.

The decision to use `infer` for static or streaming datasets is now determined by the presence of the `autoupdates` keyword. This keyword is crucial in controlling how updated posteriors are transformed into priors for the following observations, making it specifically applicable and essential for streaming data (real-world) scenarios.

**Key Changes:**

1. **Unified Inference Functionality:** The `infer` function consolidates all features previously divided between `inference` and `rxinference`. The original functions have been internally renamed to `__inference` and `__rxinference` and are no longer exported. This unified approach enables `infer` to support both batch/static and streaming/online applications.

2. **Updated Documentation:** Comprehensive updates have been made to the documentation to reflect the merging of `inference` and `rxinference`. This includes a detailed and clear docstring for the `infer` function, ensuring ease of use and understanding.

3. **Updated Examples:** All examples across the package have been updated to utilize the new `infer` function, demonstrating its application in various contexts.

4. **Deprecation Notices:** Appropriate deprecation notices have been added for the `inference` and `rxinference` functions, guiding users towards the new unified approach. (I haven't managed to make `@deprecate inference(; kwargs...) infer(kwargs...)` work, seems deprecate macro isn't good for keyword arguments)

### CODE REVIEW

**Comment by bvdmitri** [comment]
I guess we can make it mandatory and allow `missing` as an input in case someone comes up with a scenario where it is not mandatory.

**Comment by albertpod** [comment]
@bvdmitri the question is more like if we can deduct that the user wants to run `rxinference` or `inference` based on the presemse `autoupdates`, or we still better have a keyword that explicitly specifies the behaviour, i.e.,
`infer(..., setting=:filtering)`?

File: src/inference.jl [hunk 1]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
 
     return engine
 end
+
+function rxinference(; kwargs)
+    @warn "The `rxinference` function is deprecated and will be removed in the future.  Use `infer` with autoupdates keyword argument instead."
```
Comment (bvdmitri) [inline src/inference.jl:1413]:
```suggestion
    @warn "The `rxinference` function is deprecated and will be removed in the future.  Use `infer` with the `autoupdates` keyword argument instead."
```

File: src/inference.jl [hunk 2]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+- `showprogress = false`: show progress module, optional, defaults to false (exclusive for batch inference)
+- `catch_exception`  specifies whether exceptions during the inference procedure should be caught, optional, defaults to false (exclusive for batch inference)
+- `callbacks = nothing`: inference cycle callbacks, optional, see below for more info
+- `addons = nothing`: inject and send extra computation information along messages, see below for more info
+- `postprocess = DefaultPostprocess()`: inference results postprocessing step, optional, see below for more info
+- `events = nothing`: inference cycle events, optional, see below for more info (exclusive for streamline inference)
+- `uselock = false`: specifies either to use the lock structure for the inference or not, if set to true uses `Base.Threads.SpinLock`. Accepts custom `AbstractLock`. (exclusive for streamline inference)
+- `autostart = true`: specifies whether to call `RxInfer.start` on the created engine automatically or not (exclusive for streamline inference) (exclusive for streamline inference)
+- `warn = true`: enables/disables warnings
```
Comment (bvdmitri) [inline src/inference.jl:1500]: `catch_exception` argument explanation is missing (did we have it?)

File: src/inference.jl [hunk 3]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+- `showprogress = false`: show progress module, optional, defaults to false (exclusive for batch inference)
+- `catch_exception`  specifies whether exceptions during the inference procedure should be caught, optional, defaults to false (exclusive for batch inference)
+- `callbacks = nothing`: inference cycle callbacks, optional, see below for more info
+- `addons = nothing`: inject and send extra computation information along messages, see below for more info
+- `postprocess = DefaultPostprocess()`: inference results postprocessing step, optional, see below for more info
+- `events = nothing`: inference cycle events, optional, see below for more info (exclusive for streamline inference)
+- `uselock = false`: specifies either to use the lock structure for the inference or not, if set to true uses `Base.Threads.SpinLock`. Accepts custom `AbstractLock`. (exclusive for streamline inference)
+- `autostart = true`: specifies whether to call `RxInfer.start` on the created engine automatically or not (exclusive for streamline inference) (exclusive for streamline inference)
```
Comment (bvdmitri) [inline src/inference.jl:1499]:
```suggestion
- `autostart = true`: specifies whether to call `RxInfer.start` on the created engine automatically or not (exclusive for streamline inference)
```

File: src/inference.jl [hunk 4]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+    ),
+)
+
+- ### `options`
+
+- `limit_stack_depth`: limits the stack depth for computing messages, helps with `StackOverflowError` for some huge models, but reduces the performance of inference backend. Accepts integer as an argument that specifies the maximum number of recursive depth. Lower is better for stack overflow error, but worse for performance.
+- `pipeline`: changes the default pipeline for each factor node in the graph
+- `global_reactive_scheduler`: changes the scheduler of reactive streams, see Rocket.jl for more info, defaults to no scheduler
+
+- ### `returnvars`
```
Comment (bvdmitri) [inline src/inference.jl:1576]: Something has been lost here, the `returnvars` argument behaves differently for batched and streamlined versions. I think a part of the documentation has been lost in the process.

File: src/inference.jl [hunk 5]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+    predictvars = (
+        o = KeepLast(),
+        τ = KeepEach()
+    )
+)
+```
+
+**Note**: The `predictvars` argument is exclusive for batch setting.
+
+- ### `historyvars`
```
Comment (bvdmitri) [inline src/inference.jl:1612]: `historyvars` is exclusive for streamline setting

File: src/inference.jl [hunk 6]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+
+- ### `keep_history`
+
+Specifies the buffer size for the updates history both for the `historyvars` and the `free_energy` buffers in streamline inference.
+
+- ### `iterations`
+
+Specifies the number of variational (or loopy belief propagation) iterations. By default set to `nothing`, which is equivalent of doing 1 iteration. 
+
+- ### `free_energy`
```
Comment (bvdmitri) [inline src/inference.jl:1656]: The `free_energy` behaves differently for the streamlined and batched versions. In the streamlined it is indeed an observable, but in the batched version it is just a vector.

Comment (bvdmitri) [reply src/inference.jl:1656]: (again probably some documentation has been lost in the process)

File: src/inference.jl [hunk 7]
```diff
@@ -1986,3 +1408,397 @@ function rxinference(;
...
+- ### `iterations`
+
+Specifies the number of variational (or loopy belief propagation) iterations. By default set to `nothing`, which is equivalent of doing 1 iteration. 
+
+- ### `free_energy` 
+
+This setting specifies whenever the `infer` function should create an observable of Bethe Free Energy (BFE) values. The BFE observable returns a new computed value for each VMP iteration.
+Note, however, that it may be not possible to compute BFE values for every model. If `free_energy = true` and `keephistory > 0` the engine exposes extra fields to access the history of the Bethe free energy updates:
+
+- `engine.free_energy_history`: Returns a free energy history averaged over the VMP iterations
```
Comment (bvdmitri) [inline src/inference.jl:1661]: This is exclusive to the streamlined version, the batched version does not have those

**Review by bvdmitri — CHANGES_REQUESTED** [review]
Some documentation has been lost in the process. Also the function documentation is kinda big. Do we want to maybe make a documentation page instead (I'm not sure, just asking your opinion)?

File: src/inference.jl (hunk 1 above)
Comment (albertpod) [reply src/inference.jl:1413]: I suggest we create an associated issue regarding the documentation page after merging this PR.

File: src/inference.jl (hunk 2 above)
Comment (albertpod) [reply src/inference.jl:1500]: I believe the short explanation was given in line 1493, but the extended description was missing indeed. Added now.

File: src/inference.jl (hunk 4 above)
Comment (albertpod) [reply src/inference.jl:1576]: Fixed.

File: src/inference.jl (hunk 5 above)
Comment (albertpod) [reply src/inference.jl:1612]: I think it is said in both short and extended description.

File: src/inference.jl (hunk 6 above)
Comment (albertpod) [reply src/inference.jl:1656]: Fixed.

File: src/inference.jl (hunk 7 above)
Comment (albertpod) [reply src/inference.jl:1661]: Addressed

**Review by bvdmitri — APPROVED** [review]
Well done @albertpod ! You can merge after the CI pass

**Pull request merged**